src/onnx/parse_matmul.cpp

```diff
@@ -43,66 +43,127 @@ struct parse_matmul : op_parser<parse_matmul>
-        }
+            auto s0_lens = s0.lens();
+            auto s1_lens = s1.lens();
+
+            // args[0] is a vector, prepend 1 to the shape
+            bool is_a_prepended = false;
+            if(s0.ndim() == 1)
+            {
+                is_a_prepended = true;
+                s0_lens.insert(s0_lens.begin(), 1);
+                a0 = info.add_instruction(make_op("unsqueeze", {{"axes", {0}}}), args[0]);
+            }
+
+            bool is_b_appended = false;
+            if(s1.ndim() == 1)
+            {
+                is_b_appended = true;
+                s1_lens.push_back(1);
+                a1 = info.add_instruction(make_op("unsqueeze", {{"axes", {1}}}), args[1]);
+            }
 
-        return dot_res;
+            instruction_ref ba0 = a0;
+            instruction_ref ba1 = a1;
+            // try broadcasting if dimensions other than last two do not match
+            if(not std::equal(
+                   s0_lens.rbegin() + 2, s0_lens.rend(), s1_lens.rbegin() + 2, s1_lens.rend()))
+            {
+                auto l0_it = s0_lens.begin() + s0_lens.size() - 2;
+                std::vector<std::size_t> l0_broadcasted_lens(s0_lens.begin(), l0_it);
+                auto l1_it = s1_lens.begin() + s1_lens.size() - 2;
+                std::vector<std::size_t> l1_broadcasted_lens(s1_lens.begin(), l1_it);
+                auto output_lens =
+                    compute_broadcasted_lens(l0_broadcasted_lens, l1_broadcasted_lens);
+                l0_broadcasted_lens = output_lens;
+                l0_broadcasted_lens.insert(l0_broadcasted_lens.end(), l0_it, s0_lens.end());
+                l1_broadcasted_lens = output_lens;
+                l1_broadcasted_lens.insert(l1_broadcasted_lens.end(), l1_it, s1_lens.end());
+                if(s0_lens != l0_broadcasted_lens)
+                {
+                    ba0 = info.add_instruction(
+                        make_op("multibroadcast", {{"out_lens", l0_broadcasted_lens}}), a0);
+                }
+                if(s1_lens != l1_broadcasted_lens)
+                {
+                    ba1 = info.add_instruction(
+                        make_op("multibroadcast", {{"out_lens", l1_broadcasted_lens}}), a1);
+                }
+            }
+            instruction_ref dot_res = info.add_instruction(make_op(opd.op_name), ba0, ba1);
+            // squeeze the appended or prepended dimensions
+            int64_t num_axis = static_cast<int64_t>(dot_res->get_shape().lens().size());
+            if(is_a_prepended)
+            {
+                dot_res =
+                    info.add_instruction(make_op("squeeze", {{"axes", {num_axis - 2}}}), dot_res);
+                --num_axis;
+            }
+            if(is_b_appended)
+            {
+                dot_res =
+                    info.add_instruction(make_op("squeeze", {{"axes", {num_axis - 1}}}), dot_res);
+            }
+
+            return dot_res;
```

Comment:
This looks like its the repeating the same thing for dynamic and non-dynamic. This should probably be more unified.